From 0b033dd6459ddac54684d54be3fff3baf62ef106 Mon Sep 17 00:00:00 2001
From: Wyes Karny <wyes.karny@amd.com>
Date: Thu, 19 Jan 2023 17:47:49 +0000
Subject: [PATCH 18/18] acpi: cppc: Avoid writing zero value to min and max
 perf

For some cases cppc_set_perf API is called with min and max perf zero,
but on those places author didn't wanted to set those values to zero.
Before recent this API's internal update these value writes are avoided.
But after amd_pstate guided mode introduction those writes are
important. So, write those values only if those are not zero.

Signed-off-by: Wyes Karny <wyes.karny@amd.com>
---
 drivers/acpi/cppc_acpi.c | 8 ++++++--
 1 file changed, 6 insertions(+), 2 deletions(-)

diff --git a/drivers/acpi/cppc_acpi.c b/drivers/acpi/cppc_acpi.c
index d53e14de8..0fdfdc1a7 100644
--- a/drivers/acpi/cppc_acpi.c
+++ b/drivers/acpi/cppc_acpi.c
@@ -1629,8 +1629,12 @@ int cppc_set_perf(int cpu, struct cppc_perf_ctrls *perf_ctrls)
 	}
 
 	cpc_write(cpu, desired_reg, perf_ctrls->desired_perf);
-	cpc_write(cpu, min_perf_reg, perf_ctrls->min_perf);
-	cpc_write(cpu, max_perf_reg, perf_ctrls->max_perf);
+
+	/* Only write if min_perf and max_perf not zero. */
+	if (perf_ctrls->min_perf)
+		cpc_write(cpu, min_perf_reg, perf_ctrls->min_perf);
+	if (perf_ctrls->max_perf)
+		cpc_write(cpu, max_perf_reg, perf_ctrls->max_perf);
 
 	if (CPC_IN_PCC(desired_reg) || CPC_IN_PCC(min_perf_reg) || CPC_IN_PCC(max_perf_reg))
 		up_read(&pcc_ss_data->pcc_lock);	/* END Phase-I */
-- 
2.39.0.rc2.1.gbd5df96b79

